Sources/Signal.swift

```diff
@@ -1492,6 +1492,95 @@ extension SignalProtocol {
 		}
 	}
 
+	/// Conditionally throttles values sent on the receiver whenever
+	/// `shouldThrottle` is true, forwarding values on the given scheduler.
+	///
+	/// - note: While `shouldThrottle` remains false, values are forwarded on the
+	///         given scheduler.
+	///
+	/// - note: If multiple values are received while `shouldThrottle` is true,
+	///         the latest value is the one that will be passed on.
+	///
+	/// - note: If the input signal terminates while a value is being throttled,
+	///         that value will be discarded and the returned signal will
+	///         terminate immediately.
+	///
+	/// - parameters:
+	///   - shouldThrottle: A boolean property that controls whether values
+	///                     should be throttled.
+	///   - scheduler: A scheduler to deliver events on.
+	///
+	/// - returns: A signal that sends values only while `shouldThrottle` is false.
+	func throttle<P: PropertyProtocol>(while shouldThrottle: P, on scheduler: SchedulerProtocol) -> Signal<Value, Error>
+		where P.Value == Bool
+	{
+		return Signal { observer in
+			let initial: ThrottleWhileState<Value> = shouldThrottle.value ? .throttled(nil) : .resumed
+			let state = Atomic(initial)
+			let schedulerDisposable = SerialDisposable()
+
+			let disposable = CompositeDisposable()
+			disposable += schedulerDisposable
+			disposable += { _ = shouldThrottle }
+
+			disposable += shouldThrottle.producer
+				.skipRepeats()
+				.startWithValues { shouldThrottle in
+					var valueToSend: Value?
+
+					state.modify { state -> ThrottleWhileState<Value> in
+						guard !state.isTerminated else { return state }
+
+						if shouldThrottle {
+							return .throttled(nil)
+						} else {
+							if case .throttled(let value?) = state {
+								valueToSend = value
+							}
+							return .resumed
+						}
+					}
```

Comment:
The returned value from `Atomic.modify` block is not a new value for the `Atomic` instance since https://github.com/ReactiveCocoa/ReactiveCocoa/pull/2984. So this should be written as follows:

```swift
let valueToSend = state.modify { state -> Value? in
	guard !state.isTerminated else { return nil }

	if shouldThrottle {
		state = .throttled(nil)
		return nil
	} else {
		state = .resumed
		if case let .throttled(value) = state {
			return value
		} else {
			return nil
		}
	}
}
```